CI: Add bumpversion.yml workflow as a webhook target

starnight opened the pull request:
Add bumpversion.yml workflow as a webhook target/entry to bump release version. It accepts a new kolibri-explore-plugin release's trigger now. When it gets a new kolibri-explore-plugin release, it will append a dummy commit, then invoke the tag.yml workflow to create a new tag.

https://phabricator.endlessm.com/T34806

starnight committed:
CI: Add bumpversion.yml workflow as a webhook target

Add bumpversion.yml workflow as a webhook target/entry to bump release
version. It accepts a new kolibri-explore-plugin release's trigger now.
When it gets a new kolibri-explore-plugin release, it will append a
dummy commit, then invoke the tag.yml workflow to create a new tag.

https://phabricator.endlessm.com/T34806

starnight committed:
CI: Use git to create & push a new tag directly

The original action negz/create-tag@v1 using deprecated Node.js 12. So,
use git to create & push a new tag directly to avoid the warning
message.

starnight committed:
CI: Sync the tag as kolibri-explore-plugin's major & minor version

Sync the tag as kolibri-explore-plugin's major & minor version, if
kolibri-explore-plugin has a new release. And, the build number will
restart again.

Fixes: https://github.com/endlessm/endless-key-app/issues/119

manuq reviewed (approved):
Looks good to me, Nothing to comment.

dbnicholson commented on .github/workflows/bumpversion.yml:
```diff
@@ -0,0 +1,36 @@
+
+on:
+  repository_dispatch:
+    types: [kolibri-explore-plugin]
```
I think the type should probably called `kolibri-explore-plugin-release` just to be a bit more specific than "something happened on kolibri-explore-plugin".

starnight replied:
ok!

starnight replied:
Updated by the new commit.

dbnicholson commented on .github/workflows/bumpversion.yml:
```diff
@@ -0,0 +1,36 @@
+      # conflicted tag next time, due to the same tag name. So, append a dummy
+      # commit here.
+      - name: Append a new commit for new kolibri-explore-plugin
+        if: ${{ github.event.action == 'kolibri-explore-plugin' }}
```
Is `github.event.action` really going to be `kolibri-explore-plugin`? It seems like you want `github.event.event_type` per https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#repository_dispatch.

starnight replied:
I used `github.event.event_type` at first try, but it showed nothing.  Then, did some study and found it is kind of `github.event.action`.
Here is my repo's example:
* Code: https://github.com/starnight/build-image/blob/v0.0.15/.github/workflows/bumpversion.yml#L27
* Debug run: https://github.com/starnight/build-image/actions/runs/5208911111/jobs/9398076874#step:4:18

dbnicholson reviewed (changes requested):
This seems to make sense. I had a couple comments, but I want to review the kolibri-explore-plugin side, too.

starnight committed:
CI: Fix actions/checkout's ref for bumping version

The bump version feature comming from commit ("CI: Add bumpversion.yml
workflow as a webhook target") will append a new commit for a new
kolibri-explore-plugin release. So, the latest commit on master branch
will become different from github.ref. Hence, tweak actions/checkout's
ref parameter for the scenario.

https://phabricator.endlessm.com/T34806

starnight commented:
Lesson learned from https://github.com/endlessm/kolibri-explore-plugin/issues/589#issuecomment-1588459916
So, add a new commit ("CI: Fix actions/checkout's ref for bumping version").

dbnicholson commented on .github/workflows/bumpversion.yml:
```diff
@@ -0,0 +1,38 @@
+    needs: bump-version
+    uses: ./.github/workflows/tag.yml
+    with:
+      kep_verison: ${{ github.event.client_payload.VERSION }}
```
```suggestion
      kep_version: ${{ github.event.client_payload.VERSION }}
```

starnight replied:
oh!  Good catch!  thanks!

dbnicholson commented on .github/workflows/tag.yml:
```diff
@@ -3,47 +3,84 @@ name: Tag Weekly
     - cron: '32 18 * * SAT'
+  workflow_call:
+    inputs:
+      kep_verison:
```
```suggestion
      kep_version:
```

starnight replied:
oh!  Good catch!  thanks!

dbnicholson commented on .github/workflows/tag.yml:
```diff
@@ -3,47 +3,84 @@ name: Tag Weekly
           }
           echo "tag_string=$tag_str" >> $Env:GITHUB_OUTPUT
+        env:
+          kep_version: ${{ inputs.kep_verison }}
```
```suggestion
          kep_version: ${{ inputs.kep_version }}
```

starnight replied:
oh!  Good catch!  thanks!

dbnicholson commented on .github/workflows/tag.yml:
```diff
@@ -3,47 +3,84 @@ name: Tag Weekly
+
           # Do not create a new tag if latest tag is at HEAD
-          if ($revision -ne 0) {
+          if ($kle_updated -or $revision -ne 0) {
```
```suggestion
          if ($kep_updated -or $revision -ne 0) {
```

starnight replied:
oh!  Good catch!  thanks!

dbnicholson commented on .github/workflows/tag.yml:
```diff
@@ -3,47 +3,84 @@ name: Tag Weekly
-        with:
+        run: |
+          git tag -a $version -m $message
+          git push origin $version
```
I think this workflow needs the same `git config user.name` setup if it's starting from `workflow_dispatch`.

starnight replied:
ok!

dbnicholson reviewed (changes requested):
Sorry, I forgot about this PR. I just found a few typos but otherwise it looks reasonable.

dbnicholson reviewed (approved):
We'll see how it goes!

starnight commented:
[Bump_Version](https://github.com/endlessm/endless-key-app/actions/workflows/bumpversion.yml) is triggered by [Build for release / trigger_endless-key-app](https://github.com/endlessm/kolibri-explore-plugin/actions/runs/5383904247/jobs/9771639359#logs) successfully
https://github.com/endlessm/endless-key-app/actions/runs/5384111696
